docs/RationaleHWArith.md

```diff
@@ -0,0 +1,73 @@
+# HW Arith Dialect Rationale
+
+This document describes the various design points of the HWArith dialect, a dialect that is used to represent bitwidth aware arithmetic. This follows in the spirit of other
+[MLIR Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+- [HW Arith Dialect Rationale](#hw-arith-dialect-rationale)
+  - [Introduction](#introduction)
+    - [Q&A](#qa)
+  - [Bit Width Rules](#bit-width-rules)
+  - [Lowering](#lowering)
+
+## Introduction
+
+The `hwarith` dialect provides a collection of operations that define typical integer arithmetic operations which are bitwidth aware.
+This bitwidth awareness is expressed through return type inference rules that, based on the types of an operation and its input operands, infers the type of the result operand.
```

Comment:
Is this to say that it's normal operation definitions with implicit extensions?  (type inference in the IR is not the same as type inference in the language sense, it's just a shortcut for eliding redundant information). 

If you are only doing this to elide result types, that is one thing.  If you are eliding result types and allowing mismatched widths of operands, that's another point.  If you are allowing unknown width operands, that is quite another thing.

Reply:
We are both allowing mismatched operand types and doing the "real" math bit extensions on the result type.

In this first pass, we are _not_ allowing unknown widths _anywhere_. I don't think we'll be adding this anytime soon if ever.